### PR TITLE
lightgbm: migrate to Conan v2, add v4.0.0

### PR DESCRIPTION
**Pull request by valgur**
TODO:
- [ ] v1: `Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES)`

### CODE REVIEW

**Comment by valgur** [comment]
@uilianries Can you help with this one as well, please?
`ERROR: Missing prebuilt package for 'fast_double_parser/0.7.0'`

**Comment by uilianries** [comment]
@valgur I just started an internal build to generate `fast_double_parser/0.7.0`

File: recipes/lightgbm/all/conanfile.py [hunk 1]
```diff
@@ -1,82 +1,115 @@
...
+        tc.generate()
+        tc = CMakeDeps(self)
+        tc.generate()
 
     def _patch_sources(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            tools.patch(**patch)
+        cmakelists_path = os.path.join(self.source_folder, "CMakeLists.txt")
+        # Fix OpenMP detection for Clang
+        replace_in_file(self, cmakelists_path, "AppleClang", "Clang|AppleClang")
```
Comment (uilianries) [inline recipes/lightgbm/all/conanfile.py:79]: I would say it may worth a patch to the upstream, as it's small and we always need to apply it.

**Comment by uilianries** [comment]
> ERROR: Missing binary: fast_double_parser/0.7.0:da39a3ee5e6b4b0d3255bfef95601890afd80709

It's in the build queue, there are more builds running right now.

**Review by jwillikers — APPROVED** [review]

**Review by uilianries — COMMENTED** [review]
Please, revert to patches as before, so we can track that change. Plus, I opened a PR to the upstream: https://github.com/microsoft/LightGBM/pull/6178

**Pull request merged**